module/executiondatasync/tracker/storage.go

```diff
@@ -0,0 +1,454 @@
+type PruneCallback func(cid.Cid) error
+
+type Storage interface {
+	Update(UpdateFn) error
+	GetFulfilledHeight() (uint64, error)
+	SetFulfilledHeight(height uint64) error
+	GetPrunedHeight() (uint64, error)
+	Prune(height uint64) error
+}
+
+type storage struct {
+	mu            sync.RWMutex
+	db            *badger.DB
+	pruneCallback PruneCallback
+	logger        zerolog.Logger
+}
+
+type StorageOption func(*storage)
+
+func WithPruneCallback(callback PruneCallback) StorageOption {
+	return func(s *storage) {
+		s.pruneCallback = callback
+	}
+}
+
+func OpenStorage(dbPath string, startHeight uint64, logger zerolog.Logger, opts ...StorageOption) (*storage, error) {
+	db, err := badger.Open(badger.LSMOnlyOptions(dbPath))
+	if err != nil {
+		return nil, fmt.Errorf("could not open tracker db: %w", err)
+	}
+
+	storage := &storage{
+		db:            db,
+		pruneCallback: func(c cid.Cid) error { return nil },
+		logger:        logger.With().Str("module", "tracker_storage").Logger(),
+	}
+
+	for _, opt := range opts {
+		opt(storage)
+	}
+
+	if err := storage.init(startHeight); err != nil {
+		return nil, fmt.Errorf("failed to initialize storage: %w", err)
+	}
+
+	return storage, nil
+}
+
+func (s *storage) init(startHeight uint64) error {
+	fulfilledHeight, fulfilledHeightErr := s.GetFulfilledHeight()
+	prunedHeight, prunedHeightErr := s.GetPrunedHeight()
+
+	if fulfilledHeightErr == nil && prunedHeightErr == nil {
+		if prunedHeight > fulfilledHeight {
+			return fmt.Errorf("inconsistency detected: pruned height is greater than fulfilled height")
+		}
+
+		// replay pruning in case it was interrupted during previous shutdown
+		if err := s.Prune(prunedHeight); err != nil {
+			return fmt.Errorf("failed to replay pruning: %w", err)
+		}
+	} else if errors.Is(fulfilledHeightErr, badger.ErrKeyNotFound) && errors.Is(prunedHeightErr, badger.ErrKeyNotFound) {
+		// db is empty, we need to bootstrap it
+		if err := s.bootstrap(startHeight); err != nil {
+			return fmt.Errorf("failed to bootstrap storage: %w", err)
+		}
+	} else {
+		return multierror.Append(fulfilledHeightErr, prunedHeightErr).ErrorOrNil()
+	}
+
+	return nil
+}
+
+func (s *storage) bootstrap(startHeight uint64) error {
+	fulfilledHeightKey := makeGlobalStateKey(globalStateFulfilledHeight)
+	fulfilledHeightValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(fulfilledHeightValue, startHeight)
+
+	prunedHeightKey := makeGlobalStateKey(globalStatePrunedHeight)
+	prunedHeightValue := make([]byte, 8)
+	binary.LittleEndian.PutUint64(prunedHeightValue, startHeight)
+
+	return s.db.Update(func(txn *badger.Txn) error {
+		if err := txn.Set(fulfilledHeightKey, fulfilledHeightValue); err != nil {
+			return fmt.Errorf("failed to set fulfilled height value: %w", err)
+		}
+
+		if err := txn.Set(prunedHeightKey, prunedHeightValue); err != nil {
+			return fmt.Errorf("failed to set pruned height value: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (s *storage) Update(f UpdateFn) error {
```

Comment:
Please add documentation for this function. For example:
* used to add new blob CIDs to the tracker
* blob CIDs may be for either sealed or unsealed heights
* the same CID can be added multiple times, for different heights
* can the same CID be added multiple times, for the same height?